### PR TITLE
Break dependency on <nccl> inside cuda/nccl

### PR DESCRIPTION
**Pull request by thatdan23**
All CUDA NCCL functionality is implemented in https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/nccl.cpp which is part of libtorch_cuda
https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/python_nccl.cpp should be a thin wrapper that should not use any symbols from NCCL library nor depend on any header.
Currently it includes nccl.h for a few enums, defines, which makes guaranteed that libtorch_python should not use any NCCL symbols hard to enforce.

This PR seeks to break that dependency in the header, so as to limit it's impact.

### CODE REVIEW

File: torch/csrc/cuda/nccl.h [hunk 1]
```diff
@@ -6,11 +6,18 @@
...
 
-#include <nccl.h>
-
 #include <cstddef>
 #include <vector>
 
+enum ncclDataType_t : unsigned int;
+enum ncclResult_t : unsigned int;
+enum ncclRedOp_t : unsigned int;
+struct ncclComm_t;
```
Comment (malfet) [inline torch/csrc/cuda/nccl.h:15]: This forward declaration is wrong, isn't it?
At least according to https://github.com/NVIDIA/nccl/blob/5949d96f36d050e59d05872f8bbffd2549318e95/src/nccl.h.in#L26 it should be something like
```suggestion
typedef struct ncclComm* ncclComm_t;
```

File: torch/csrc/cuda/nccl.h [hunk 2]
```diff
@@ -6,11 +6,18 @@
...
 #include <vector>
 
+enum ncclDataType_t : unsigned int;
+enum ncclResult_t : unsigned int;
+enum ncclRedOp_t : unsigned int;
+typedef struct ncclComm* ncclComm_t;
+struct ncclUniqueId;
+int32_t ncclSum;
+
+
```
Comment (malfet) [inline torch/csrc/cuda/nccl.h:19]: Why `ncclSum` global variable is declared in this header? Also wouldn't it lead to a linker conflict?
```suggestion
int32_t ncclSum;


```

Comment (thatdan23) [reply torch/csrc/cuda/nccl.h:19]: ncclSum is one of the values in ncclRedOp_t.  I suppose I could replace the default with a hard coded value, I assume ncclRedOp_t::ncclSum is 0.

**Review by malfet — APPROVED** [review]

File: torch/csrc/cuda/nccl.h [hunk 3]
```diff
@@ -6,55 +6,87 @@
...
 namespace cuda {
 namespace nccl {
 
+    /*
+The below enums are from
+https://github.com/NVIDIA/nccl/blob/master/src/nccl.h.in
+however they're declare anyonomously there, so we can't forward declare them
+here.*/
+/* Error type */
+typedef enum {
```
Comment (malfet) [inline torch/csrc/cuda/nccl.h:24]: Nit: it would be nice to have those as properly namespace scoped `enum classes` rather than global C style enums

Comment (thatdan23) [reply torch/csrc/cuda/nccl.h]: fixed

**Pull request closed without merging**